1 wired system

yuriykuzin opened the pull request.

yuriykuzin committed:
Update README.md

yuriykuzin committed:
Add files via upload

yuriykuzin committed:
Update 1-wiredsystem.html

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+  function Node() {
+    var children = [];    
+    
+    this.howManyChildren = function() {
```
Лучше `childrenLength` или `childrenCount`, и круто было бы сделать его таким же свойством, как собственно array.length

yuriykuzin replied:
Ок, делаю так:

```
  this.childrenCount = 0;
  var self = this;
...
  self.childrenCount++;   // внутри метода attach

```

Но! В глобальной перспективе (если не ограничиваться рамками задачи) получается доступное вовне свойство, которое можно менять извне объекта, но при этом оно используется во внутренней логике объекта. Поэтому мне показалось, что лучше геттер использовать. Или есть еще какие-то варианты?

albert-tarly replied:
Есть :) Посмотри в сторону Object.defineProperty

yuriykuzin replied:
Ага!

```
Object.defineProperty(this, "childrenCount", {
    get: function() {      
      return children.length;
    }
  });  
```

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+    };
+    
+    this.attach = function(nodeToAttach) {    
+      if (!(nodeToAttach instanceof Node)) throw new TypeError("Can only attach nodes");
```
Плохой стиль писать условие в одну строку, легко потерять вторую. Лучше всегда писать фигурные скобки.

yuriykuzin replied:
Roger that, thanks!

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+  LampNode.prototype = Object.create(Node.prototype);
+
+  function SwitchNode(signal) {
+    Node.call(this);
```
То есть у свитча будет метод attach, чилдрены и вот это всё? :)

yuriykuzin replied:
Верно, убираю Node.call(this);

albert-tarly replied:
Хитро, но как-то опасно)

yuriykuzin replied:
Опасно - если Node в перспективе будет расширяться новыми методами?
Какие есть альтернативы? Что я думаю:
- можно переопределить метод attach в SwitchNode, но children все равно останутся в багажнике.
- можно ввести промежуточный класс NodeWithChildren, и только в нем создавать attach, childrenCount, children.

Что-то еще?

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+    
+    var parentAttach = this.attach;      
+    this.attach = function(newNode) {
+      if (this.howManyChildren() == 2) throw new RangeError("Cannot attach more than 2 nodes to this gate");
```
Нестрогое сравнение без нужды лучше не использовать

yuriykuzin replied:
Понял!

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  GateNotNode.prototype = Object.create(GateNode.prototype);
+
```
А где OR? :)

yuriykuzin replied:
В тестах не значился :)  Добавляю:

```
function GateOrNode() {
  GateNode.call(this);
  this.valueOf = function() {       
    return (this.child(0).valueOf() | this.child(1).valueOf());
  }; 
}
GateOrNode.prototype = Object.create(GateNode.prototype);
```

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+    GateNode.call(this);
+
+    this.valueOf = function() {  	       
+      return +!(this.child(0).valueOf());
```
Скобки лишние, а за `+` респект, я об этом забыл

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+    GateNode.call(this);
+
+    this.valueOf = function() {  	
+      return (this.child(0).valueOf() & this.child(1).valueOf());
```
Кстати, а почему тут нету +? :)

yuriykuzin replied:
Так ведь здесь побитовая операция, а не логическая. На выход идет число, а не Boolean

albert-tarly replied:
туплю \* 2

albert-tarly commented on 1-wiredsystem.html:
```diff
@@ -0,0 +1,151 @@
+  <meta charset="utf-8">
+</head>
+<body>
+  <script>
```
Давай вынесем в отдельный файл, править джаваскрипт в html не годится

yuriykuzin replied:
Done!

yuriykuzin committed:
Add files via upload

yuriykuzin committed:
Update 1-wiredsystem.html

yuriykuzin committed:
Update 1-wiredsystem.js

yuriykuzin committed:
With prototype declaration and inheritance

- rewrited with fully prototype declaration and inheritance;
- NodeCanHaveChild class is inserted.

yuriykuzin committed:
upd for better readability and expandability 

- put prototype and prototype.constructor settings into function inherit(son, father);
- call -> apply, so we can add more arguments later without changing this line in constructors.
